src/drivers/intel/cavs/mn.c

```diff
@@ -142,73 +171,190 @@ static bool find_mn(uint32_t freq, uint32_t bclk,
-		/* select M/N output for bclk in case of Audio Cardinal
-		 * or PLL Fixed clock.
-		 */
-		if (ssp_freq_sources[clk_index] != SSP_CLOCK_XTAL_OSCILLATOR)
-			*out_need_ecs = true;
-	} else {
-		/* check if we can get target BCLK with M/N */
-		for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++) {
-			if (find_mn(ssp_freq[i].freq, bclk_rate,
-				    out_scr_div, &i2s_m, &i2s_n)) {
-				clk_index = i;
-				break;
-			}
-		}
+	bclk_source_mn_clock = clk_index;
+
+	mn_reg_write(0x0, (mn_reg_read(0x0) |
+			   MNDSS(ssp_freq_sources[clk_index])));
+	return 0;
+}
+
+/**
+ * \brief Finds valid M/(N * SCR) values for source clock that is already locked
+ *	  because other ports use it.
+ * \param[in] bclk Bit clock frequency.
+ * \param[out] scr_div SCR divisor.
+ * \param[out] m M value of M/N divider.
+ * \param[out] n N value of M/N divider.
+ * \return 0 on success, error code otherwise.
+ */
+static inline int setup_current_bclk_mn_source(uint32_t bclk, uint32_t *scr_div,
+					       uint32_t *m, uint32_t *n)
+{
+	/* source for M/N is already set, no need to do it */
+	if (find_mn(ssp_freq[bclk_source_mn_clock].freq, bclk, scr_div, m, n))
+		return 0;
+
+	trace_mn_error("error: BCLK %d, no valid configuration for already selected source = %d",
+		       bclk, bclk_source_mn_clock);
+	return -EINVAL;
+}
 
-		if (clk_index < 0) {
-			trace_mn_error("error: BCLK %d", bclk_rate);
-			ret = -EINVAL;
-			goto out;
+#if CAVS_VERSION >= CAVS_VERSION_2_0
+static inline bool check_bclk_xtal_source(uint32_t bclk, bool mn_in_use,
+					  uint32_t *scr_div)
+{
+	/* since cAVS 2.0 bypassing XTAL (ECS=0) is not supported */
+	return false;
+}
+#else
+/**
+ * \brief Checks if XTAL source for BCLK should be used.
+ *	  Before cAVS 2.0 BCLK could use XTAL directly (without M/N).
+ *	  BCLK that use M/N = 1/1 or bypass XTAL is preferred.
+ * \param[in] bclk Bit clock frequency.
+ * \param[in] mn_in_use True if M/N source is already locked by another port.
+ * \param[out] scr_div SCR divisor.
+ * \return true if XTAL source should be used, false otherwise.
+ */
+static inline bool check_bclk_xtal_source(uint32_t bclk, bool mn_in_use,
+					  uint32_t *scr_div)
+{
+	int i;
+
+	for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++) {
+		if (ssp_freq[i].freq % bclk != 0)
+			continue;
+
+		if (ssp_freq_sources[i] == SSP_CLOCK_XTAL_OSCILLATOR) {
+			/* XTAL turned out to be lowest source that can work
+			 * just with SCR, so use it
+			 */
+			*scr_div = ssp_freq[i].freq / bclk;
+			return true;
 		}
 
-		trace_mn("M = %d, N = %d", i2s_m, i2s_n);
+		/* check if can be handled in M/N setup */
+		if (!mn_in_use || bclk_source_mn_clock == i)
+			return false;
+	}
+
+	return false;
+}
+#endif
+
+int mn_set_bclk(uint32_t dai_index, uint32_t bclk_rate,
+		uint32_t *out_scr_div, bool *out_need_ecs)
+{
+	uint32_t m = 1;
+	uint32_t n = 1;
+	int ret = 0;
+	bool mn_in_use;
+
+	spin_lock(mn_lock);
 
-		mdivc |= MNDSS(ssp_freq_sources[clk_index]);
+	mn_in_use = is_bclk_source_in_use(MN_BCLK_SOURCE_MN);
```

Comment:
@jajanusz do we need this function? Can we init bclk_source_mn_clock to -1 in mn_init() and check if it is not negative?

Reply:
@ranj063 You have multiple MCLKs (on our devices I think we have 2 everywhere), with negative/nonnegative check you don't know if user count is f.e. 1 or 2. Alternative to the table used here is ref counting (-/+ on get/put), but it's more error prone if you do 2 config and reconfig f.e. without releasing it in-between.

Reply:
Changed to bools to not cause confusion with unnecessary enum